libnavigation-core/src/main/java/com/mapbox/navigation/core/telemetry/MapboxNavigationTelemetry.kt

```diff
@@ -130,695 +103,378 @@ internal object MapboxNavigationTelemetry : MapboxNavigationTelemetryInterface {
+        log("resetOriginalRoute")
+        originalRoute = null
     }
 
-    /**
-     * This method waits for an [RouteProgressState.ROUTE_COMPLETE] event. Once received, it terminates the wait-loop and
-     * sends the telemetry data to the servers.
-     */
-    private suspend fun monitorSession() {
-        var continueRunning = true
-        var trackingEvent = 0
-        while (coroutineContext.isActive && continueRunning) {
-            try {
-                val routeData = callbackDispatcher.getRouteProgressChannel().receive()
-                dynamicValues.distanceCompleted.set(
-                    dynamicValues.distanceCompleted.get() +
-                        routeData.routeProgress.distanceTraveled
-                )
-                dynamicValues.distanceRemaining.set(
-                    routeData.routeProgress.distanceRemaining.toLong()
-                )
-                dynamicValues.durationRemaining.set(
-                    routeData.routeProgress.durationRemaining.toInt()
-                )
-                when (routeData.routeProgress.currentState) {
-                    RouteProgressState.ROUTE_COMPLETE -> {
-                        when (dynamicValues.sessionStarted.get()) {
-                            true -> {
-                                processArrival()
-                                continueRunning = false
-                            }
-                            false -> {
-                                // Do nothing.
-                                Log.d(TAG, "route arrival received before a session start")
-                            }
-                        }
-                    } // END
-                    RouteProgressState.LOCATION_TRACKING -> {
-                        dynamicValues.timeRemaining.set(
-                            callbackDispatcher
-                                .getRouteProgress()
-                                .routeProgress
-                                .durationRemaining
-                                .toInt()
-                        )
-                        dynamicValues.distanceRemaining.set(
-                            callbackDispatcher
-                                .getRouteProgress()
-                                .routeProgress
-                                .distanceRemaining
-                                .toLong()
-                        )
-                        when (trackingEvent > 20) {
-                            true -> {
-                                Log.i(TAG, "LOCATION_TRACKING received $trackingEvent")
-                                trackingEvent = 0
-                            }
-                            false -> {
-                                trackingEvent++
-                            }
-                        }
-                    }
-                    else -> {
-                        // Do nothing
+    private var needStartSession = false
+
+    private fun startSessionIfNeedAndCan() {
```

Comment:
This is critical business logic. Are we covering the different use cases with tests?

Reply:
yep, it's covered, have added more